File: steganography_cli/engine/crypto.py
"""
Carnation Radio - Encryption Layer

Provides AES-256-GCM encryption for the steganography payload.
Key is derived from a user passphrase using scrypt (memory-hard KDF).

Usage:
    from crypto import encrypt_message, decrypt_message

    ciphertext = encrypt_message(b"secret message", passphrase="my password")
    plaintext = decrypt_message(ciphertext, passphrase="my password")

Wire format of encrypted payload:
    [salt:16][nonce:16][tag:16][ciphertext:N]
    Total overhead: 48 bytes
"""

import os

from Crypto.Cipher import AES
from Crypto.Protocol.KDF import scrypt


SALT_SIZE = 16
NONCE_SIZE = 16  # AES-GCM recommended nonce size
TAG_SIZE = 16
SCRYPT_N = 2**14  # CPU/memory cost (lower for speed; increase for production)
SCRYPT_R = 8
SCRYPT_P = 1
KEY_SIZE = 32  # AES-256


def derive_key(passphrase: str, salt: bytes) -> bytes:
    """Derive a 256-bit key from passphrase using scrypt."""
    return scrypt(passphrase.encode("utf-8"), salt, KEY_SIZE,
                  N=SCRYPT_N, r=SCRYPT_R, p=SCRYPT_P)


def encrypt_message(plaintext: bytes, passphrase: str) -> bytes:
    """
    Encrypt a message with AES-256-GCM using a passphrase.

    Returns: salt + nonce + tag + ciphertext
    """
    salt = os.urandom(SALT_SIZE)
    key = derive_key(passphrase, salt)

    cipher = AES.new(key, AES.MODE_GCM, nonce=os.urandom(NONCE_SIZE))
    ciphertext, tag = cipher.encrypt_and_digest(plaintext)

    return salt + cipher.nonce + tag + ciphertext


def decrypt_message(payload: bytes, passphrase: str) -> bytes:
    """
    Decrypt a message encrypted with encrypt_message().

    Raises ValueError if passphrase is wrong or data is corrupted.
    """
    min_size = SALT_SIZE + NONCE_SIZE + TAG_SIZE
    if len(payload) < min_size:
        raise ValueError(f"Payload too short: {len(payload)} < {min_size}")

    salt = payload[:SALT_SIZE]
    nonce = payload[SALT_SIZE:SALT_SIZE + NONCE_SIZE]
    tag = payload[SALT_SIZE + NONCE_SIZE:SALT_SIZE + NONCE_SIZE + TAG_SIZE]
    ciphertext = payload[SALT_SIZE + NONCE_SIZE + TAG_SIZE:]

    key = derive_key(passphrase, salt)
    cipher = AES.new(key, AES.MODE_GCM, nonce=nonce)

    try:
        return cipher.decrypt_and_verify(ciphertext, tag)
    except ValueError:
        raise ValueError("Decryption failed - wrong passphrase or corrupted data")

File: frontend/lib/crypto.ts
import { scrypt } from '@noble/hashes/scrypt.js'

const SALT_SIZE = 16
const NONCE_SIZE = 16
const TAG_SIZE = 16
const SCRYPT_N = 2 ** 14
const SCRYPT_R = 8
const SCRYPT_P = 1
const KEY_SIZE = 32

export function deriveKey(passphrase: string, salt: Uint8Array): Uint8Array {
  return scrypt(new TextEncoder().encode(passphrase), salt, {
    N: SCRYPT_N, r: SCRYPT_R, p: SCRYPT_P, dkLen: KEY_SIZE,
  })
}

export async function encryptMessage(plaintext: Uint8Array, passphrase: string): Promise<Uint8Array> {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_SIZE))
  const keyBytes = deriveKey(passphrase, salt)
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_SIZE))

  const key = await crypto.subtle.importKey('raw', keyBytes, 'AES-GCM', false, ['encrypt'])
  const encrypted = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce, tagLength: 128 },
    key, plaintext,
  )
  const encArr = new Uint8Array(encrypted)

  // Web Crypto appends tag. Rearrange to: salt + nonce + tag + ciphertext
  const ciphertext = encArr.slice(0, encArr.length - TAG_SIZE)
  const tag = encArr.slice(encArr.length - TAG_SIZE)

  const result = new Uint8Array(SALT_SIZE + NONCE_SIZE + TAG_SIZE + ciphertext.length)
  result.set(salt, 0)
  result.set(nonce, SALT_SIZE)
  result.set(tag, SALT_SIZE + NONCE_SIZE)
  result.set(ciphertext, SALT_SIZE + NONCE_SIZE + TAG_SIZE)
  return result
}

export async function decryptMessage(payload: Uint8Array, passphrase: string): Promise<Uint8Array> {
  if (payload.length < SALT_SIZE + NONCE_SIZE + TAG_SIZE)
    throw new Error('Payload too short')

  const salt = payload.slice(0, SALT_SIZE)
  const nonce = payload.slice(SALT_SIZE, SALT_SIZE + NONCE_SIZE)
  const tag = payload.slice(SALT_SIZE + NONCE_SIZE, SALT_SIZE + NONCE_SIZE + TAG_SIZE)
  const ciphertext = payload.slice(SALT_SIZE + NONCE_SIZE + TAG_SIZE)

  const keyBytes = deriveKey(passphrase, salt)
  const key = await crypto.subtle.importKey('raw', keyBytes, 'AES-GCM', false, ['decrypt'])

  // Web Crypto expects tag appended to ciphertext
  const combined = new Uint8Array(ciphertext.length + TAG_SIZE)
  combined.set(ciphertext)
  combined.set(tag, ciphertext.length)

  const decrypted = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: nonce, tagLength: 128 },
    key, combined,
  )
  return new Uint8Array(decrypted)
}
